crates/bevy_core/src/time/time.rs

```diff
@@ -1,95 +1,264 @@
-use bevy_ecs::system::ResMut;
+use bevy_ecs::system::{Res, ResMut};
 use bevy_utils::{Duration, Instant};
 
-/// Tracks elapsed time since the last update and since the App has started
+use crate::{FixedTime, FixedTimestepState};
+
+/// `Time` tracks how much time has advanced (and also the raw CPU time elapsed) since its previous update and since the app was started.
 #[derive(Debug, Clone)]
 pub struct Time {
-    delta: Duration,
+    startup: Instant,
+    first_update: Option<Instant>,
     last_update: Option<Instant>,
-    delta_seconds_f64: f64,
+    relative_speed: f64, // using `f64` instead of `f32` to minimize drift from rounding errors
+    delta: Duration,
     delta_seconds: f32,
-    seconds_since_startup: f64,
-    time_since_startup: Duration,
-    startup: Instant,
+    delta_seconds_f64: f64,
+    elapsed_since_startup: Duration,
+    seconds_since_startup: f32,
+    seconds_since_startup_f64: f64,
+    raw_delta: Duration,
+    raw_delta_seconds: f32,
+    raw_delta_seconds_f64: f64,
+    raw_elapsed_since_startup: Duration,
+    raw_seconds_since_startup: f32,
+    raw_seconds_since_startup_f64: f64,
 }
 
 impl Default for Time {
-    fn default() -> Time {
-        Time {
-            delta: Duration::from_secs(0),
-            last_update: None,
+    fn default() -> Self {
+        Self {
             startup: Instant::now(),
+            first_update: None,
+            last_update: None,
+            relative_speed: 1.0,
+            delta: Duration::ZERO,
+            delta_seconds: 0.0,
             delta_seconds_f64: 0.0,
+            elapsed_since_startup: Duration::ZERO,
             seconds_since_startup: 0.0,
-            time_since_startup: Duration::from_secs(0),
-            delta_seconds: 0.0,
+            seconds_since_startup_f64: 0.0,
+            raw_delta: Duration::ZERO,
+            raw_delta_seconds: 0.0,
+            raw_delta_seconds_f64: 0.0,
+            raw_elapsed_since_startup: Duration::ZERO,
+            raw_seconds_since_startup: 0.0,
+            raw_seconds_since_startup_f64: 0.0,
         }
     }
 }
 
 impl Time {
+    /// Constructs a new `Time` instance with a specific startup `Instant`.
+    pub fn new(startup: Instant) -> Self {
+        Self {
+            startup,
+            ..Default::default()
+        }
+    }
+
     /// Updates the internal time measurements.
     pub fn update(&mut self) {
-        self.update_with_instant(Instant::now());
+        let now = Instant::now();
+        self.update_with_instant(now);
     }
 
+    /// Advances time and records the [`Instant`] it happened.
     pub(crate) fn update_with_instant(&mut self, instant: Instant) {
         if let Some(last_update) = self.last_update {
```

Comment:
There's a few parts in there that are identical in both branches. I think it would be easier to see what's actually different if it was extracted after the if/else.

Reply:
I see what you mean, I've updated it and moved the common parts outside.